src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/StackItem.cs

```diff
@@ -3,68 +3,117 @@
+        private long _int64;
+
+        [FieldOffset(8)]
+        private IntPtr _nativeInt;
+
+        [FieldOffset(8)]
+        private float _float;
+
+        [FieldOffset(8)]
+        private double _double;
 
-        public StackItem(T value, StackValueKind kind, WellKnownType type)
+        [FieldOffset(16)]
+        private ValueType _valueType;
+
+        [FieldOffset(16)]
+        private object _objref;
+
+        public StackValueKind Kind => _kind;
+
+        public static StackItem FromInt32(int int32)
         {
-            Value = value;
-            Kind = kind;
-            Type = type;
+            return new StackItem { _int32 = int32, _kind = StackValueKind.Int32 };
         }
-    }
 
-    internal class Int32StackItem : StackItem<int>
-    {
-        public Int32StackItem(int value) : base(value, StackValueKind.Int32, WellKnownType.Int32)
+        public int AsInt32()
         {
+            Debug.Assert(_kind == StackValueKind.Int32);
+            return _int32;
         }
-    }
 
-    internal class Int64StackItem : StackItem<long>
-    {
-        public Int64StackItem(long value) : base(value, StackValueKind.Int64, WellKnownType.Int64)
+        public static StackItem FromInt64(long int64)
         {
+            return new StackItem { _int64 = int64, _kind = StackValueKind.Int64 };
         }
-    }
 
-    internal class FloatStackItem : StackItem<float>
-    {
-        public FloatStackItem(float value) : base(value, StackValueKind.Float, WellKnownType.Single)
+        public long AsInt64()
         {
+            Debug.Assert(_kind == StackValueKind.Int64);
+            return _int64;
         }
-    }
 
-    internal class DoubleStackItem : StackItem<double>
-    {
-        public DoubleStackItem(double value) : base(value, StackValueKind.Float, WellKnownType.Double)
+        public static StackItem FromIntPtr(IntPtr nativeInt)
         {
+            return new StackItem { _nativeInt = nativeInt, _kind = StackValueKind.NativeInt };
         }
-    }
 
-    internal class ValueTypeStackItem : StackItem<ValueType>
-    {
-        public ValueTypeStackItem(ValueType value) : base(value, StackValueKind.ValueType, WellKnownType.ValueType)
+        public IntPtr AsIntPtr()
         {
+            Debug.Assert(_kind == StackValueKind.NativeInt);
+            return _nativeInt;
         }
-    }
 
-    internal class ObjectRefStackItem : StackItem<Object>
-    {
-        public ObjectRefStackItem(Object value) : base(value, StackValueKind.ObjRef, WellKnownType.Object)
+        public static StackItem FromFloat(float single)
+        {
+            return new StackItem { _float = single, _kind = StackValueKind.Float };
+        }
+
+        public float AsFloat()
+        {
+            Debug.Assert(_kind == StackValueKind.Float);
+            return _float;
+        }
+
+        public static StackItem FromDouble(double d)
+        {
+            return new StackItem { _double = d, _kind = StackValueKind.Float };
```

Comment:
You won't be able to tell whether the `_double` is valid or `_float` is valid since both of them use same `StackValueKind.Float`. 

There are two options:
- Have 64-bit float only in the StackItem and StackValueKind. It is acceptable behavior according to the IL spec.
- Have both 32-bit float and 64-bit float in both the StackItem and StackValueKind.